### ISSUE TITLE
`--s3-directory-markers` flag  and not working `rclone purge`

### ISSUE THREAD

**Issue by kapitainsky** [rclone/rclone#7247]
When`--s3-directory-markers` flag is set for S3 compatible remote it is possible to simulate empty "directories" by uploading an empty object with a trailing slash when a new directory is created. This works perfectly fine.

The problem is when attempting to delete such "directory".  It is impossible.

Remote:

```
[iDrive]
type = s3
provider = IDrive
access_key_id = XXX
secret_access_key = XXX
endpoint = XXX
directory_markers = true
```

```
$ rclone mkdir iDrive:test2/test

$ rclone lsf iDrive:test2
test/

$ rclone purge iDrive:test2/test -vv
2023/08/24 15:03:56 DEBUG : rclone: Version "v1.63.1" starting with parameters ["rclone" "purge" "iDrive:test2/test" "-vv"]
2023/08/24 15:03:56 DEBUG : Creating backend with remote "iDrive:test2/test"
2023/08/24 15:03:56 DEBUG : Using config file from "/Users/kptsky/.config/rclone/rclone.conf"
2023/08/24 15:03:56 DEBUG : name = "iDrive", root = "test2/test", opt = &s3.Options{Provider:"IDrive", EnvAuth:false, AccessKeyID:"XXX", SecretAccessKey:"XXX", Region:"", Endpoint:"p7v1.ldn.idrivee2-40.com", STSEndpoint:"", LocationConstraint:"", ACL:"", BucketACL:"", RequesterPays:false, ServerSideEncryption:"", SSEKMSKeyID:"", SSECustomerAlgorithm:"", SSECustomerKey:"", SSECustomerKeyBase64:"", SSECustomerKeyMD5:"", StorageClass:"", UploadCutoff:209715200, CopyCutoff:4999341932, ChunkSize:5242880, MaxUploadParts:10000, DisableChecksum:false, SharedCredentialsFile:"", Profile:"", SessionToken:"", UploadConcurrency:4, ForcePathStyle:true, V2Auth:false, UseAccelerateEndpoint:false, LeavePartsOnError:false, ListChunk:1000, ListVersion:0, ListURLEncode:fs.Tristate{Value:false, Valid:false}, NoCheckBucket:false, NoHead:false, NoHeadObject:false, Enc:0x3000002, MemoryPoolFlushTime:60000000000, MemoryPoolUseMmap:false, DisableHTTP2:false, DownloadURL:"", DirectoryMarkers:true, UseMultipartEtag:fs.Tristate{Value:false, Valid:false}, UsePresignedRequest:false, Versions:false, VersionAt:fs.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}, Decompress:false, MightGzip:fs.Tristate{Value:false, Valid:false}, UseAcceptEncodingGzip:fs.Tristate{Value:false, Valid:false}, NoSystemMetadata:false}
2023/08/24 15:03:56 DEBUG : Resolving service "s3" region "us-east-1"
2023/08/24 15:03:56 DEBUG : S3 bucket test2 path test: bucket is versioned: false
2023/08/24 15:03:56 DEBUG : Waiting for deletions to finish
2023/08/24 15:03:56 DEBUG : "" version false
2023/08/24 15:03:56 DEBUG : : Deleting (id "")
2023/08/24 15:03:57 INFO  : : Deleted
2023/08/24 15:03:57 DEBUG : 7 go routines active

$ rclone lsf  iDrive:test2
test/
```

I have tested it with AWS and iDrive remotes.

**Comment by ncw** [comment]
Is it just the top level directory that isn't being deleted?

**Comment by kapitainsky** [comment]
> Is it just the top level directory that isn't being deleted?

Nope - the bucket/top level dir in this example is test2.

The `rclone purge` is supposed to delete directory and all its content. 

With `directory_markers = true` it fail regardless if it is empty or contains any content.

**Comment by kapitainsky** [comment]
> Is it just the top level directory that isn't being deleted?

When I create subfolders and upload some files that indeed it is only top level which is not deleted.

**Comment by ncw** [comment]
This turned out to be quite a subtle bug complicated by the fact that S3 doesn't give an error if you attempt to delete a non-existent file!

Anyway, give this a go

[v1.64.0-beta.7291.b95bda1e9.fix-7247-s3-purge-root](https://beta.rclone.org/branch/fix-7247-s3-purge-root/v1.64.0-beta.7291.b95bda1e9.fix-7247-s3-purge-root/) on branch [fix-7247-s3-purge-root](https://github.com/rclone/rclone/tree/fix-7247-s3-purge-root) (uploaded in 15-30 mins)

**Comment by kapitainsky** [comment]
I have tested this fix and now `rclone purge` behaves as expected deleting all folders.

thank you very much.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.64

Thanks for testing :-)
